libc/stdio/stdio_indicator.c

```diff
@@ -0,0 +1,628 @@
+
+		TEST_ASSERT_EQUAL_INT64_MESSAGE(0, ftell(f), tellMode(modes[i]));
+	}
+	TEST_ASSERT_EQUAL_INT(0, remove(filename));
+}
+
+
+TEST(stdio_ftell, bad_file_descriptor)
+{
+	errno = 0;
+	char filename[] = "test_bfd.txt";
+	FILE *f = fopen(filename, "w+");
+	TEST_ASSERT_EQUAL_INT(0, fclose(f));
+	TEST_ASSERT_EQUAL(-1, ftell(f));
+	TEST_ASSERT_EQUAL_INT(EBADF, errno);
+	TEST_ASSERT_EQUAL_INT(0, remove(filename));
+}
+
+
+TEST(stdio_ftell, wrong_stream_type)
+{
+	TEST_IGNORE_MESSAGE("#923 issue");
+
+
+	int socketfd = socket(AF_UNIX, SOCK_STREAM, 0);
+
+	TEST_ASSERT_NOT_EQUAL_INT(-1, socketfd);
+
+	errno = 0;
+	FILE *socket_stream = fdopen(socketfd, "w+");
+
+	TEST_ASSERT_NOT_NULL_MESSAGE(socket_stream, strerror(errno));
+
+	errno = 0;
+	TEST_ASSERT_EQUAL_INT64(-1, ftell(socket_stream));
+	TEST_ASSERT_EQUAL_INT(ESPIPE, errno);
+
+	errno = 0;
+	TEST_ASSERT_EQUAL_INT64(-1, ftello(socket_stream));
+	TEST_ASSERT_EQUAL_INT(ESPIPE, errno);
+
+	TEST_ASSERT_EQUAL_INT(0, fclose(socket_stream));
+	TEST_ASSERT_NOT_EQUAL_INT(EOF, close(socketfd));
+
+
+	errno = 0;
+	int pipefd[2];
+	int pipe_res = pipe(pipefd);
+
+	TEST_ASSERT_NOT_EQUAL_INT(-1, pipe_res);
+
+	FILE *pipe_stream = fdopen(pipefd[0], "r");
+	TEST_ASSERT_NOT_NULL(pipe_stream);
+
+	errno = 0;
+	TEST_ASSERT_EQUAL_INT64(-1, ftell(pipe_stream));
+	TEST_ASSERT_EQUAL_INT(ESPIPE, errno);
+
+	errno = 0;
+	TEST_ASSERT_EQUAL_INT64(-1, ftello(pipe_stream));
+	TEST_ASSERT_EQUAL_INT(ESPIPE, errno);
+
+	TEST_ASSERT_EQUAL_INT(0, fclose(pipe_stream));
+	TEST_ASSERT_NOT_EQUAL_INT(EOF, close(pipefd[0]));
+	TEST_ASSERT_NOT_EQUAL_INT(EOF, close(pipefd[1]));
+
+
+	const char *fifo_path = "myfifo";
+	TEST_ASSERT_NOT_EQUAL_INT(-1, mkfifo(fifo_path, 0666));
+
+	pid_t pid = fork();
+	TEST_ASSERT_NOT_EQUAL_INT(-1, pid);
+
+	if (!pid) {
+		int fifofd = open(fifo_path, O_WRONLY);
+
+		if (fifofd == -1) {
+			exit(EXIT_FAILURE);
+		}
+
+		if (write(fifofd, "1234", 4) == -1) {
+			close(fifofd);
+			exit(EXIT_FAILURE);
+		}
+
+		if (close(fifofd) == EOF) {
+			exit(EXIT_FAILURE);
+		}
+
+		exit(EXIT_SUCCESS);
+	}
+	else {
+		int status;
+		wait(&status);
+		TEST_ASSERT_EQUAL_INT(EXIT_SUCCESS, status);
+
+		int fifofd = open(fifo_path, O_RDWR);
+
+		TEST_ASSERT_NOT_EQUAL(-1, fifofd);
+
+		FILE *fifo_stream = fdopen(fifofd, "w+");
+
+		TEST_ASSERT_NOT_NULL(fifo_stream);
+
+		TEST_ASSERT_NOT_EQUAL(EOF, fflush(fifo_stream));
+
+		errno = 0;
+		TEST_ASSERT_EQUAL(-1, ftell(fifo_stream));
+		TEST_ASSERT_EQUAL(ESPIPE, errno);
+
+		errno = 0;
+		TEST_ASSERT_EQUAL(-1, ftello(fifo_stream));
+		TEST_ASSERT_EQUAL(ESPIPE, errno);
+
+
+		TEST_ASSERT_EQUAL_INT(0, fclose(fifo_stream));
+		TEST_ASSERT_NOT_EQUAL(EOF, close(fifofd));
+	}
+
+
+	TEST_ASSERT_EQUAL(0, remove(fifo_path));
+}
+
+
+TEST_GROUP_RUNNER(stdio_ftell)
+{
+	RUN_TEST_CASE(stdio_ftell, correct_position_not_empty);
+	RUN_TEST_CASE(stdio_ftell, correct_position_empty);
+	RUN_TEST_CASE(stdio_ftell, bad_file_descriptor);
+	RUN_TEST_CASE(stdio_ftell, wrong_stream_type);
```

Comment:
I think that `EOVERFLOW` errno may be also checked.

Reply:
We can prepare a few edge cases, so when file content equals long max, long max - 1 and long max + 1, same for off_t.

Reply:
Maybe some separate testcase for checking this reuirement? (or checking it everywhere else where ftell call is successful)
`The ftell() function shall not change the setting of errno if successful.`